---
bool any_frontiers_left